File: src/Library/Calculator/ExpTree/Nodes/Collections/EnumerableCollectionNode.cs
﻿using System.Collections.Generic;

namespace Calculator.ExpTree.Nodes.Collections
{
    public abstract class EnumerableCollectionNode : BranchNode
    {
        protected List<ExpNode> Children { get; }

        public override int ChildCount => Children.Count;

        public override Priority Priority => Priority.OVERRIDE;

        public override void AddChild(ExpNode node)
        {
            Children.Add(node);
        }

        public override void InsertChild(BranchNode node)
        {
            ExpNode lastChild = Children[ChildCount - 1];
            Children.RemoveAt(ChildCount - 1);
            node.AddChild(lastChild);
            AddChild(node);
        }

        public override ExpNode GetChild(int index)
        {
            return Children[index];
        }

        public override void ClearChildren()
        {
            Children.Clear();
        }
    }
}

File: old/C++/src/Library/Calculator/EnumerableCollectionNode.cpp
#include "EnumerableCollectionNode.h"

/// <summary>
/// Adds a child and sets its parent
/// </summary>
/// <param name="node">New child node</param>
void EnumerableCollectionNode::AddChild(unique_ptr<ExpNode> node) {
	children_.push_back(move(node));
}

/// <summary>
/// Inserts child between this and its last child
/// </summary>
/// <param name="node">this's new child node</param>
void EnumerableCollectionNode::InsertChild(unique_ptr<BranchNode> node) {
	// Gets this's last child and remove from children_
	node->AddChild(move(children_.back()));
	children_.pop_back();
	AddChild(move(node));
}

/// <summary>
/// Gets child at index
/// </summary>
/// <returns>child at index</returns>
const ExpNode& EnumerableCollectionNode::GetChild(int index) const {
	return *children_[index];
}

void EnumerableCollectionNode::ClearChildren() {
	children_.clear();
}

/// <summary>
/// Gets amount of children
/// </summary>
/// <returns>children count</returns>
int EnumerableCollectionNode::ChildCount() const {
	return children_.size();
}

/// <summary>
/// Gets the priority based on the oper
/// </summary>
/// <returns>Proirity value</returns>
Priority EnumerableCollectionNode::GetPriority() const {
	return Priority::VALUE;
}

bool EnumerableCollectionNode::IsConstant() const {
	for (size_t i = 0; i < children_.size(); i++)
	{
		if (!children_[i]->IsConstant()) {
			return false;
		}
	}
	return true;
}

bool EnumerableCollectionNode::IsConstantBy(const VarValueNode& node) const {
	for (size_t i = 0; i < children_.size(); i++)
	{
		if (!children_[i]->IsConstantBy(node)) {
			return false;
		}
	}
	return true;
}